File: client/src/app/core/models/requests/bulk-update-tasks-request.ts
import { BulkCollectionMode } from '@core/enums/bulk-collection-mode';
import { EstimateType } from '@core/enums/estimate-type';
import { TaskPriority } from '@core/enums/task-priority';

// Each field is applied only when provided; sprint and due date use their clear flags to remove a
// value. Tags and assignees carry a mode saying whether they replace what a task holds or join it.
export interface BulkUpdateTasksRequest {
  taskIds: number[];
  statusId?: number | null;
  priority?: TaskPriority | null;
  estimateType?: EstimateType | null;
  estimateValue?: number | null;
  projectId?: number | null;
  sprintId?: number | null;
  clearSprint?: boolean;
  dueDate?: string | null;
  clearDueDate?: boolean;
  assigneeIds?: string[] | null;
  assigneeMode?: BulkCollectionMode;
  tags?: string[] | null;
  tagMode?: BulkCollectionMode;
}

File: server/Netptune.Core/Requests/BulkUpdateTasksRequest.cs
using Netptune.Core.Enums;

namespace Netptune.Core.Requests;

// Each field is "set if provided, otherwise leave unchanged". Sprint and due date have explicit
// Clear flags so they can be removed (null can't distinguish "clear" from "keep"). Tags and
// assignees carry a mode, because a bulk edit is as often "add these" as "make it exactly these".
public class BulkUpdateTasksRequest
{
    public List<int> TaskIds { get; init; } = [];

    public int? StatusId { get; init; }

    public TaskPriority? Priority { get; init; }

    public EstimateType? EstimateType { get; init; }

    public decimal? EstimateValue { get; init; }

    public int? ProjectId { get; init; }

    public int? SprintId { get; init; }

    public bool ClearSprint { get; init; }

    public DateOnly? DueDate { get; init; }

    public bool ClearDueDate { get; init; }

    public List<string>? AssigneeIds { get; init; }

    public BulkCollectionMode AssigneeMode { get; init; }

    public List<string>? Tags { get; init; }

    public BulkCollectionMode TagMode { get; init; }
}
